File: backend/app/services/analytics_service.py
"""
Analytics Service - Microservice for real-time metrics and reports
Part of Phase 3: Support Services Architecture

Enhanced with:
- Real-time analytics processing
- Advanced data aggregation
- Predictive analytics capabilities
- Cross-service metrics collection
- Performance monitoring
- Business intelligence reporting
"""

import logging
from datetime import date, datetime, timedelta
from decimal import Decimal
from enum import Enum

from sqlmodel import Session, and_, func, select

from .analytics import OrderAnalytics, RevenueAnalytics

from app.models import (
    Cart,
    Customer,
    OrderStatus,
    Product,
    SalesItem,
    SalesOrder,
    Stock,
    Transaction,
    TransactionType,
)

logger = logging.getLogger(__name__)


class MetricType(str, Enum):
    """Analytics metric types"""

    REVENUE = "REVENUE"
    ORDERS = "ORDERS"
    CUSTOMERS = "CUSTOMERS"
    PRODUCTS = "PRODUCTS"
    INVENTORY = "INVENTORY"
    CONVERSION = "CONVERSION"
    PERFORMANCE = "PERFORMANCE"
    ENGAGEMENT = "ENGAGEMENT"


class TimeGranularity(str, Enum):
    """Time granularity for analytics"""

    MINUTE = "MINUTE"
    HOUR = "HOUR"
    DAY = "DAY"
    WEEK = "WEEK"
    MONTH = "MONTH"
    QUARTER = "QUARTER"
    YEAR = "YEAR"


class AnalyticsService:
    """Enhanced service for calculating business analytics and KPIs with microservices support"""

    def __init__(self, db: Session):
        self.db = db
        self._metrics_cache = {}
        self._real_time_buffer = []
        self._revenue = RevenueAnalytics(db)
        self._orders = OrderAnalytics(db)

    # ─── REVENUE METRICS ──────────────────────────────────────────────────────

    def get_total_revenue(
        self, start_date: datetime | None = None, end_date: datetime | None = None
    ) -> Decimal:
        """Calculate total revenue for a given period"""
        return self._revenue.get_total_revenue(start_date, end_date)

    def get_daily_revenue(self, target_date: date) -> Decimal:
        """Calculate revenue for a specific day"""
        return self._revenue.get_daily_revenue(target_date)

    def get_monthly_revenue(self, year: int, month: int) -> Decimal:
        """Calculate revenue for a specific month"""
        return self._revenue.get_monthly_revenue(year, month)

    def get_revenue_growth_rate(self, period_days: int = 30) -> float:
        """Calculate revenue growth rate comparing two periods"""
        return self._revenue.get_revenue_growth_rate(period_days)

    def calculate_mrr(self) -> Decimal:
        """Calculate Monthly Recurring Revenue (MRR)"""
        return self._revenue.calculate_mrr()

    def calculate_arr(self) -> Decimal:
        """Calculate Annual Recurring Revenue (ARR)"""
        return self._revenue.calculate_arr()

    def calculate_cac(self, channel: str = None) -> Decimal:
        """Calculate Customer Acquisition Cost (CAC)"""
        return self._revenue.calculate_cac(channel)

    def calculate_revenue_per_visitor(self, days: int = 30) -> Decimal:
        """Calculate Revenue Per Visitor (RPV)"""
        return self._revenue.calculate_revenue_per_visitor(days)

    # ─── ORDER METRICS ────────────────────────────────────────────────────────

    def get_average_order_value(
        self, start_date: datetime | None = None, end_date: datetime | None = None
    ) -> Decimal:
        """Calculate Average Order Value (AOV)"""
        return self._orders.get_average_order_value(start_date, end_date)

    def get_order_metrics(self) -> dict:
        """Get comprehensive order metrics"""
        return self._orders.get_order_metrics()

    # ─── CUSTOMER METRICS ─────────────────────────────────────────────────────

    def get_customer_metrics(self) -> dict:
        """Get comprehensive customer metrics"""
        # Total customers
        total_customers = (
            self.db.exec(select(func.count(Customer.customer_id))).first() or 0
        )

        # New customers this month
        month_start = datetime(datetime.utcnow().year, datetime.utcnow().month, 1)
        new_customers_month = (
            self.db.exec(
                select(func.count(Customer.customer_id)).where(
                    Customer.created_at >= month_start
                )
            ).first()
            or 0
        )

        # Customers with orders
        customers_with_orders = (
            self.db.exec(
                select(func.count(func.distinct(SalesOrder.customer_id))).where(
                    SalesOrder.status == OrderStatus.COMPLETED
                )
            ).first()
            or 0
        )

        # Active customers (ordered in last 30 days)
        thirty_days_ago = datetime.utcnow() - timedelta(days=30)
        active_customers = (
            self.db.exec(
                select(func.count(func.distinct(SalesOrder.customer_id))).where(
                    and_(
                        SalesOrder.order_date >= thirty_days_ago,
                        SalesOrder.status == OrderStatus.COMPLETED,
                    )
                )
            ).first()
            or 0
        )

        return {
            "total_customers": total_customers,
            "new_customers_month": new_customers_month,
            "customers_with_orders": customers_with_orders,
            "active_customers_30d": active_customers,
            "customer_conversion_rate": round(
                (
                    (customers_with_orders / total_customers * 100)
                    if total_customers > 0
                    else 0
                ),
                2,
            ),
        }

    def calculate_customer_lifetime_value(self, customer_id: int) -> Decimal:
        """Calculate CLV for a specific customer"""
        # Total revenue from customer
        total_spent = self.db.exec(
            select(func.sum(Transaction.amount)).where(
                and_(
                    Transaction.customer_id == customer_id,
                    Transaction.tx_type == TransactionType.SALE,
                    Transaction.paid,
                )
            )
        ).first() or Decimal(0)

        # Number of orders
        order_count = (
            self.db.exec(
                select(func.count(SalesOrder.so_id)).where(
                    and_(
                        SalesOrder.customer_id == customer_id,
                        SalesOrder.status == OrderStatus.COMPLETED,
                    )
                )
            ).first()
            or 0
        )

        # Customer lifespan in days
        customer = self.db.get(Customer, customer_id)
        if not customer:
            return Decimal(0)

        lifespan_days = (datetime.utcnow() - customer.created_at).days
        if lifespan_days == 0:
            lifespan_days = 1

        # Simple CLV calculation
        avg_order_value = total_spent / order_count if order_count > 0 else Decimal(0)
        purchase_frequency = (
            order_count / (lifespan_days / 30) if lifespan_days > 0 else 0
        )

        # Assume 24 month customer lifespan and 20% profit margin
        clv = avg_order_value * purchase_frequency * 24 * Decimal(0.2)

        return round(clv, 2)

    # ─── PRODUCT METRICS ──────────────────────────────────────────────────────

    def get_top_selling_products(self, limit: int = 10) -> list[dict]:
        """Get top selling products by quantity"""
        query = (
            select(
                Product.product_id,
                Product.name,
                Product.sku,
                func.sum(SalesItem.qty).label("total_sold"),
                func.sum(SalesItem.qty * SalesItem.unit_price).label("total_revenue"),
            )
            .join(SalesItem)
            .join(SalesOrder)
            .where(SalesOrder.status == OrderStatus.COMPLETED)
            .group_by(Product.product_id, Product.name, Product.sku)
            .order_by(func.sum(SalesItem.qty).desc())
            .limit(limit)
        )

        results = self.db.exec(query).all()

        return [
            {
                "product_id": row.product_id,
                "name": row.name,
                "sku": row.sku,
                "total_sold": row.total_sold,
                "total_revenue": float(row.total_revenue),
            }
            for row in results
        ]

    def get_inventory_metrics(self) -> dict:
        """Get inventory-related metrics"""
        # Total products
        total_products = (
            self.db.exec(select(func.count(Product.product_id))).first() or 0
        )

        # Low stock products (quantity < 10)
        low_stock_products = (
            self.db.exec(
                select(func.count(Stock.stock_id)).where(Stock.quantity < 10)
            ).first()
            or 0
        )

        # Out of stock products
        out_of_stock = (
            self.db.exec(
                select(func.count(Stock.stock_id)).where(Stock.quantity == 0)
            ).first()
            or 0
        )

        # Total inventory value
        total_inventory_value = self.db.exec(
            select(func.sum(Stock.quantity * Product.unit_price)).join(Product)
        ).first() or Decimal(0)

        return {
            "total_products": total_products,
            "low_stock_products": low_stock_products,
            "out_of_stock_products": out_of_stock,
            "total_inventory_value": float(total_inventory_value),
        }

    # ─── CONVERSION METRICS ───────────────────────────────────────────────────

    def get_cart_abandonment_rate(self, days: int = 30) -> float:
        """Calculate cart abandonment rate"""
        start_date = datetime.utcnow() - timedelta(days=days)

        # Total carts created
        total_carts = (
            self.db.exec(
                select(func.count(Cart.cart_id)).where(Cart.created_at >= start_date)
            ).first()
            or 0
        )

        # Completed orders (carts that converted)
        completed_orders = (
            self.db.exec(
                select(func.count(SalesOrder.so_id)).where(
                    and_(
                        SalesOrder.order_date >= start_date,
                        SalesOrder.status == OrderStatus.COMPLETED,
                    )
                )
            ).first()
            or 0
        )

        if total_carts == 0:
            return 0.0

        abandonment_rate = ((total_carts - completed_orders) / total_carts) * 100
        return round(abandonment_rate, 2)

    def calculate_conversion_rate(self, days: int = 30) -> float:
        """Calculate overall conversion rate"""
        start_date = datetime.utcnow() - timedelta(days=days)

        # Total unique customers who visited/interacted
        total_customers = (
            self.db.exec(
                select(func.count(func.distinct(Cart.customer_id))).where(
                    Cart.created_at >= start_date
                )
            ).first()
            or 0
        )

        # Customers who completed orders
        converting_customers = (
            self.db.exec(
                select(func.count(func.distinct(SalesOrder.customer_id))).where(
                    and_(
                        SalesOrder.order_date >= start_date,
                        SalesOrder.status == OrderStatus.COMPLETED,
                    )
                )
            ).first()
            or 0
        )

        if total_customers == 0:
            return 0.0

        conversion_rate = (converting_customers / total_customers) * 100
        return round(conversion_rate, 2)

    def calculate_repeat_customer_rate(self, days: int = 30) -> float:
        """Calculate repeat customer rate"""
        start_date = datetime.utcnow() - timedelta(days=days)

        # Customers with multiple orders
        repeat_customers = self.db.exec(
            select(func.count(func.distinct(SalesOrder.customer_id)))
            .where(SalesOrder.order_date >= start_date)
            .group_by(SalesOrder.customer_id)
            .having(func.count(SalesOrder.so_id) > 1)
        ).all()

        # Total customers who made orders
        total_customers = (
            self.db.exec(
                select(func.count(func.distinct(SalesOrder.customer_id))).where(
                    SalesOrder.order_date >= start_date
                )
            ).first()
            or 0
        )

        if total_customers == 0:
            return 0.0

        repeat_rate = (len(repeat_customers) / total_customers) * 100
        return round(repeat_rate, 2)

    def calculate_churn_rate(self, period_days: int = 90) -> float:
        """Calculate customer churn rate"""
        start_period = datetime.utcnow() - timedelta(days=period_days)

        # Customers who were active in the period
        active_customers = (
            self.db.exec(
                select(func.count(func.distinct(SalesOrder.customer_id))).where(
                    SalesOrder.order_date >= start_period
                )
            ).first()
            or 0
        )

        # Customers who haven't ordered in the last 30 days
        recent_cutoff = datetime.utcnow() - timedelta(days=30)
        churned_customers = (
            self.db.exec(
                select(func.count(func.distinct(SalesOrder.customer_id))).where(
                    and_(
                        SalesOrder.order_date >= start_period,
                        SalesOrder.customer_id.not_in(
                            select(func.distinct(SalesOrder.customer_id)).where(
                                SalesOrder.order_date >= recent_cutoff
                            )
                        ),
                    )
                )
            ).first()
            or 0
        )

        if active_customers == 0:
            return 0.0

        churn_rate = (churned_customers / active_customers) * 100
        return round(churn_rate, 2)

    # ─── FINANCIAL METRICS ────────────────────────────────────────────────────

    def get_financial_summary(self) -> dict:
        """Get comprehensive financial summary"""
        today = datetime.utcnow().date()
        month_start = datetime(today.year, today.month, 1)
        year_start = datetime(today.year, 1, 1)

        return {
            "revenue": {
                "today": float(self.get_daily_revenue(today)),
                "month": float(self.get_total_revenue(month_start)),
                "year": float(self.get_total_revenue(year_start)),
                "growth_rate": self.get_revenue_growth_rate(),
                "mrr": float(self.calculate_mrr()),
                "arr": float(self.calculate_arr()),
                "revenue_per_visitor": float(self.calculate_revenue_per_visitor()),
            },
            "orders": self.get_order_metrics(),
            "customers": self.get_customer_metrics(),
            "inventory": self.get_inventory_metrics(),
            "conversion": {
                "cart_abandonment_rate": self.get_cart_abandonment_rate(),
                "conversion_rate": self.calculate_conversion_rate(),
                "repeat_customer_rate": self.calculate_repeat_customer_rate(),
                "churn_rate": self.calculate_churn_rate(),
            },
        }

    # ─── REAL-TIME METRICS ────────────────────────────────────────────────────

    def get_realtime_metrics(self) -> dict:
        """Get real-time metrics for dashboard"""
        return {
            "current_revenue_today": float(
                self.get_daily_revenue(datetime.utcnow().date())
            ),
            "orders_today": self.get_order_metrics()["orders_today"],
            "pending_orders": self.get_order_metrics()["pending_orders"],
            "active_carts": self.db.exec(
                select(func.count(Cart.cart_id)).where(
                    Cart.updated_at >= datetime.utcnow() - timedelta(hours=24)
                )
            ).first()
            or 0,
            "low_stock_alerts": self.get_inventory_metrics()["low_stock_products"],
            "timestamp": datetime.utcnow().isoformat(),
        }

File: frontend/src/services/AnalyticsService.ts
// AnalyticsService.ts - Frontend service for analytics API calls

export interface FinancialSummary {
  revenue: {
    today: number
    month: number
    year: number
    growth_rate: number
    mrr: number
    arr: number
    revenue_per_visitor: number
  }
  orders: {
    orders_today: number
    orders_month: number
    pending_orders: number
    average_order_value: number
  }
  customers: {
    total_customers: number
    new_customers_month: number
    customers_with_orders: number
    active_customers_30d: number
    customer_conversion_rate: number
  }
  inventory: {
    total_products: number
    low_stock_products: number
    out_of_stock_products: number
    total_inventory_value: number
  }
  conversion: {
    cart_abandonment_rate: number
    conversion_rate: number
    repeat_customer_rate: number
    churn_rate: number
  }
}

export interface RealtimeMetrics {
  current_revenue_today: number
  orders_today: number
  pending_orders: number
  active_carts: number
  low_stock_alerts: number
  timestamp: string
}

export interface Alert {
  id: string
  type: string
  severity: "info" | "warning" | "critical"
  title: string
  description: string
  data: Record<string, any>
  timestamp: string
  created_at: string
}

export interface AlertSummary {
  total_alerts: number
  critical_alerts: number
  warning_alerts: number
  info_alerts: number
  last_checked: string
  alerts: Alert[]
}

export interface ApiResponse<T> {
  data: T
  success: boolean
  error?: string
}

class AnalyticsService {
  private static readonly BASE_URL = "/api/analytics"

  private static async makeRequest<T>(
    endpoint: string,
  ): Promise<ApiResponse<T>> {
    try {
      const token = localStorage.getItem("access_token")
      if (!token) {
        throw new Error("No authentication token found")
      }

      const response = await fetch(`${AnalyticsService.BASE_URL}${endpoint}`, {
        method: "GET",
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json",
        },
      })

      if (!response.ok) {
        const errorText = await response.text()
        throw new Error(`HTTP ${response.status}: ${errorText}`)
      }

      const data = await response.json()
      return { data, success: true }
    } catch (error) {
      console.error(`Analytics API Error [${endpoint}]:`, error)
      return {
        data: null as T,
        success: false,
        error: error instanceof Error ? error.message : "Unknown error",
      }
    }
  }

  /**
   * Get comprehensive financial summary
   */
  static async getFinancialSummary(): Promise<ApiResponse<FinancialSummary>> {
    return AnalyticsService.makeRequest<FinancialSummary>("/financial-summary")
  }

  /**
   * Get real-time metrics
   */
  static async getRealtimeMetrics(): Promise<ApiResponse<RealtimeMetrics>> {
    return AnalyticsService.makeRequest<RealtimeMetrics>("/realtime-metrics")
  }

  /**
   * Get alert summary
   */
  static async getAlertSummary(): Promise<ApiResponse<AlertSummary>> {
    return AnalyticsService.makeRequest<AlertSummary>("/alerts/summary")
  }

  /**
   * Get all current alerts
   */
  static async getAllAlerts(): Promise<
    ApiResponse<{ alerts: Alert[]; total_count: number; timestamp: string }>
  > {
    return AnalyticsService.makeRequest("/alerts/all")
  }

  /**
   * Get Monthly Recurring Revenue
   */
  static async getMRR(): Promise<
    ApiResponse<{ mrr: number; currency: string; calculation_date: string }>
  > {
    return AnalyticsService.makeRequest("/revenue/mrr")
  }

  /**
   * Get Annual Recurring Revenue
   */
  static async getARR(): Promise<
    ApiResponse<{ arr: number; currency: string; calculation_date: string }>
  > {
    return AnalyticsService.makeRequest("/revenue/arr")
  }

  /**
   * Get conversion rate
   */
  static async getConversionRate(days = 30): Promise<
    ApiResponse<{
      conversion_rate_percentage: number
      analysis_period_days: number
      calculation_date: string
    }>
  > {
    return AnalyticsService.makeRequest(`/conversion/rate?days=${days}`)
  }

  /**
   * Get repeat customer rate
   */
  static async getRepeatCustomerRate(days = 30): Promise<
    ApiResponse<{
      repeat_customer_rate_percentage: number
      analysis_period_days: number
      calculation_date: string
    }>
  > {
    return AnalyticsService.makeRequest(`/customers/repeat-rate?days=${days}`)
  }

  /**
   * Get churn rate
   */
  static async getChurnRate(periodDays = 90): Promise<
    ApiResponse<{
      churn_rate_percentage: number
      retention_rate_percentage: number
      analysis_period_days: number
      calculation_date: string
    }>
  > {
    return AnalyticsService.makeRequest(
      `/customers/churn-rate?period_days=${periodDays}`,
    )
  }

  /**
   * Get revenue per visitor
   */
  static async getRevenuePerVisitor(days = 30): Promise<
    ApiResponse<{
      revenue_per_visitor: number
      currency: string
      analysis_period_days: number
      calculation_date: string
    }>
  > {
    return AnalyticsService.makeRequest(`/revenue/per-visitor?days=${days}`)
  }

  /**
   * Get revenue growth rate
   */
  static async getRevenueGrowthRate(periodDays = 30): Promise<
    ApiResponse<{
      growth_rate_percentage: number
      period_days: number
      comparison_periods: string
    }>
  > {
    return AnalyticsService.makeRequest(
      `/revenue/growth-rate?period_days=${periodDays}`,
    )
  }

  /**
   * Get top selling products
   */
  static async getTopSellingProducts(limit = 10): Promise<
    ApiResponse<{
      top_products: Array<{
        product_id: number
        name: string
        sku: string
        total_sold: number
        total_revenue: number
      }>
      limit: number
      total_returned: number
    }>
  > {
    return AnalyticsService.makeRequest(`/products/top-selling?limit=${limit}`)
  }

  /**
   * Mock data for development/fallback
   */
  static getMockFinancialSummary(): FinancialSummary {
    return {
      revenue: {
        today: 2500.0,
        month: 45000.0,
        year: 350000.0,
        growth_rate: 15.2,
        mrr: 38000.0,
        arr: 456000.0,
        revenue_per_visitor: 12.5,
      },
      orders: {
        orders_today: 12,
        orders_month: 180,
        pending_orders: 8,
        average_order_value: 85.5,
      },
      customers: {
        total_customers: 1250,
        new_customers_month: 95,
        customers_with_orders: 780,
        active_customers_30d: 320,
        customer_conversion_rate: 62.4,
      },
      inventory: {
        total_products: 45,
        low_stock_products: 3,
        out_of_stock_products: 1,
        total_inventory_value: 125000.0,
      },
      conversion: {
        cart_abandonment_rate: 28.5,
        conversion_rate: 3.2,
        repeat_customer_rate: 42.1,
        churn_rate: 8.7,
      },
    }
  }

  static getMockRealtimeMetrics(): RealtimeMetrics {
    return {
      current_revenue_today: 2847.5,
      orders_today: 14,
      pending_orders: 6,
      active_carts: 23,
      low_stock_alerts: 3,
      timestamp: new Date().toISOString(),
    }
  }

  static getMockAlertSummary(): AlertSummary {
    return {
      total_alerts: 3,
      critical_alerts: 1,
      warning_alerts: 2,
      info_alerts: 0,
      last_checked: new Date().toISOString(),
      alerts: [
        {
          id: "inventory_out_of_stock_1",
          type: "inventory_out_of_stock",
          severity: "critical",
          title: "Out of Stock Alert - 1 Product",
          description: "Whey Protein Gold Standard is out of stock",
          data: { affected_products: 1 },
          timestamp: new Date().toISOString(),
          created_at: new Date().toISOString(),
        },
        {
          id: "inventory_low_stock_2",
          type: "inventory_low_stock",
          severity: "warning",
          title: "Low Stock Alert - 3 Products",
          description: "Multiple products are running low on stock",
          data: { affected_products: 3 },
          timestamp: new Date().toISOString(),
          created_at: new Date().toISOString(),
        },
      ],
    }
  }

  /**
   * Get financial summary with fallback
   */
  static async getFinancialSummaryWithFallback(): Promise<FinancialSummary> {
    const response = await AnalyticsService.getFinancialSummary()

    if (response.success && response.data) {
      return response.data
    }

    console.warn(
      "Failed to fetch financial summary, using mock data:",
      response.error,
    )
    return AnalyticsService.getMockFinancialSummary()
  }

  /**
   * Get realtime metrics with fallback
   */
  static async getRealtimeMetricsWithFallback(): Promise<RealtimeMetrics> {
    const response = await AnalyticsService.getRealtimeMetrics()

    if (response.success && response.data) {
      return response.data
    }

    console.warn(
      "Failed to fetch realtime metrics, using mock data:",
      response.error,
    )
    return AnalyticsService.getMockRealtimeMetrics()
  }

  /**
   * Get alert summary with fallback
   */
  static async getAlertSummaryWithFallback(): Promise<AlertSummary> {
    const response = await AnalyticsService.getAlertSummary()

    if (response.success && response.data) {
      return response.data
    }

    console.warn("Failed to fetch alerts, using mock data:", response.error)
    return AnalyticsService.getMockAlertSummary()
  }
}

export default AnalyticsService
